File: core/src/main/java/com/hxw/core/utils/EncryptUtils.java
package com.hxw.core.utils;


import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.spec.AlgorithmParameterSpec;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.DESKeySpec;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import androidx.annotation.Nullable;

/**
 * 加密相关工具类
 *
 * @author hxw
 * @date 2018/5/7
 */
public final class EncryptUtils {
//    private static final char[] HEX_DIGITS =
//            {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    private EncryptUtils() {
    }

    /**
     * Return the hex string of MD5 encryption.
     *
     * @param data The data.
     * @return the hex string of MD5 encryption
     */
    public static String encryptMD5ToString(@Nullable byte[] data) {
        return HexUtils.bytes2HexStr2(encryptMD5(data));
    }

    /**
     * Return the bytes of MD5 encryption.
     *
     * @param data The data.
     * @return the bytes of MD5 encryption
     */
    public static byte[] encryptMD5(@Nullable byte[] data) {
        return hashTemplate(data, "MD5");
    }

    /**
     * Return the hex string of SHA1 encryption.
     *
     * @param data The data.
     * @return the hex string of SHA1 encryption
     */
    public static String encryptSHA1ToString(@Nullable byte[] data) {
        return HexUtils.bytes2HexStr2(encryptSHA1(data));
    }

    /**
     * Return the bytes of SHA1 encryption.
     *
     * @param data The data.
     * @return the bytes of SHA1 encryption
     */
    public static byte[] encryptSHA1(@Nullable byte[] data) {
        return hashTemplate(data, "SHA1");
    }

    /**
     * Return the bytes of hash encryption.
     *
     * @param data      The data.
     * @param algorithm The name of hash encryption.
     * @return the bytes of hash encryption
     */
    private static byte[] hashTemplate(@Nullable byte[] data, final String algorithm) {
        if (data == null || data.length == 0) {
            return new byte[0];
        }
        try {
            MessageDigest md = MessageDigest.getInstance(algorithm);
            md.update(data);
            return md.digest();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return new byte[0];
        }
    }

    /**
     * Return the hex string of AES encryption.
     *
     * @param data           The data.
     * @param key            The key.
     * @param transformation The name of the transformation, e.g., <i>DES/CBC/PKCS5Padding</i><i>AES/ECB/NoPadding</i>.
     * @param initVector     The buffer with the IV. The contents of the
     *                       buffer are copied to protect against subsequent modification.
     * @return the hex string of AES encryption
     */
    public static String encryptAES2HexString(byte[] data,
                                              byte[] key,
                                              String transformation,
                                              byte[] initVector) {
        return HexUtils.bytes2HexStr1(encryptAES(data, key, transformation, initVector));
    }

    /**
     * Return the bytes of AES decryption for hex string.
     *
     * @param data           The data.
     * @param key            The key.
     * @param transformation The name of the transformation, e.g., <i>DES/CBC/PKCS5Padding</i><i>AES/ECB/NoPadding</i>.
     * @param iv             The buffer with the IV. The contents of the
     *                       buffer are copied to protect against subsequent modification.
     * @return the bytes of AES decryption for hex string
     */
    public static byte[] decryptHexStringAES(String data,
                                             byte[] key,
                                             String transformation,
                                             byte[] iv) {
        return decryptAES(HexUtils.hexStr2Bytes(data), key, transformation, iv);
    }


    /**
     * Return the bytes of AES encryption.
     *
     * @param data           The data.
     * @param key            The key.
     * @param transformation The name of the transformation, e.g., <i>DES/CBC/PKCS5Padding</i><i>AES/ECB/NoPadding</i>.
     * @param initVector     The buffer with the IV. The contents of the
     *                       buffer are copied to protect against subsequent modification.
     * @return the bytes of AES encryption
     */
    public static byte[] encryptAES(byte[] data,
                                    byte[] key,
                                    String transformation,
                                    byte[] initVector) {
        return symmetricTemplate(data, key, "AES", transformation, initVector, true);
    }

    /**
     * Return the bytes of AES decryption.
     *
     * @param data           The data.
     * @param key            The key.
     * @param transformation The name of the transformation, e.g., <i>DES/CBC/PKCS5Padding</i><i>AES/ECB/NoPadding</i>.
     * @param iv             The buffer with the IV. The contents of the
     *                       buffer are copied to protect against subsequent modification.
     * @return the bytes of AES decryption
     */
    public static byte[] decryptAES(byte[] data,
                                    byte[] key,
                                    String transformation,
                                    byte[] iv) {
        return symmetricTemplate(data, key, "AES", transformation, iv, false);
    }

    /**
     * Return the bytes of symmetric encryption or decryption.
     *
     * @param data           The data.
     * @param key            The key.
     * @param algorithm      The name of algorithm.
     * @param transformation The name of the transformation, e.g., <i>DES/CBC/PKCS5Padding</i><i>AES/ECB/NoPadding</i>.
     * @param isEncrypt      True to encrypt, false otherwise.
     * @return the bytes of symmetric encryption or decryption
     */
    private static byte[] symmetricTemplate(@Nullable byte[] data,
                                            @Nullable byte[] key,
                                            String algorithm,
                                            String transformation,
                                            byte[] initVector,
                                            boolean isEncrypt) {
        if (data == null || key == null || data.length == 0 || key.length == 0) {
            return new byte[0];
        }
        try {
            SecretKey secretKey;
            if ("DES".equals(algorithm)) {
                DESKeySpec desKey = new DESKeySpec(key);
                SecretKeyFactory keyFactory = SecretKeyFactory.getInstance(algorithm);
                secretKey = keyFactory.generateSecret(desKey);
            } else {
                secretKey = new SecretKeySpec(key, algorithm);
            }
            Cipher cipher = Cipher.getInstance(transformation);
            if (initVector == null || initVector.length == 0) {
                cipher.init(isEncrypt ? Cipher.ENCRYPT_MODE : Cipher.DECRYPT_MODE, secretKey);
            } else {
                AlgorithmParameterSpec params = new IvParameterSpec(initVector);
                cipher.init(isEncrypt ? Cipher.ENCRYPT_MODE : Cipher.DECRYPT_MODE, secretKey, params);
            }
            return cipher.doFinal(data);
        } catch (Throwable e) {
            e.printStackTrace();
            return new byte[0];
        }
    }
}

File: core/src/main/java/com/hxw/core/utils/EncryptUtils.kt
package com.hxw.core.utils


import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.spec.AlgorithmParameterSpec

import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.DESKeySpec
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * 加密相关工具类
 *
 * @author hxw
 * @date 2018/5/7
 */

//    private static final char[] HEX_DIGITS =
//            {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};


/**
 * Return the hex string of MD5 encryption.
 *
 * @param data The data.
 * @return the hex string of MD5 encryption
 */
fun encryptMD5ToString(data: ByteArray?): String {
    return if (data == null || data.isEmpty()) {
        ""
    } else HexUtils.bytes2HexStr2(encryptMD5(data))
}

/**
 * Return the bytes of MD5 encryption.
 *
 * @param data The data.
 * @return the bytes of MD5 encryption
 */
fun encryptMD5(data: ByteArray): ByteArray {
    return hashTemplate(data, "MD5")
}

/**
 * Return the hex string of SHA1 encryption.
 *
 * @param data The data.
 * @return the hex string of SHA1 encryption
 */
fun encryptSHA1ToString(data: ByteArray?): String {
    return if (data == null || data.isEmpty()) {
        ""
    } else HexUtils.bytes2HexStr2(encryptSHA1(data))
}

/**
 * Return the bytes of SHA1 encryption.
 *
 * @param data The data.
 * @return the bytes of SHA1 encryption
 */
fun encryptSHA1(data: ByteArray): ByteArray {
    return hashTemplate(data, "SHA1")
}

/**
 * Return the bytes of hash encryption.
 *
 * @param data      The data.
 * @param algorithm The name of hash encryption.
 * @return the bytes of hash encryption
 */
private fun hashTemplate(data: ByteArray, algorithm: String): ByteArray {
    return try {
        val md = MessageDigest.getInstance(algorithm)
        md.update(data)
        md.digest()
    } catch (e: NoSuchAlgorithmException) {
        e.printStackTrace()
        ByteArray(0)
    }
}

/**
 * Return the hex string of AES encryption.
 *
 * @param data           The data.
 * @param key            The key.
 * @param transformation The name of the transformation, e.g., *DES/CBC/PKCS5Padding*.
 * @param initVector     The buffer with the IV. The contents of the
 * buffer are copied to protect against subsequent modification.
 * @return the hex string of AES encryption
 */
fun encryptAES2HexString(
    data: ByteArray,
    key: ByteArray,
    transformation: String,
    initVector: ByteArray
): String {
    return HexUtils.bytes2HexStr1(encryptAES(data, key, transformation, initVector))
}

/**
 * Return the bytes of AES encryption.
 *
 * @param data           The data.
 * @param key            The key.
 * @param transformation The name of the transformation, e.g., *DES/CBC/PKCS5Padding*.
 * @param initVector     The buffer with the IV. The contents of the
 * buffer are copied to protect against subsequent modification.
 * @return the bytes of AES encryption
 */
fun encryptAES(
    data: ByteArray,
    key: ByteArray,
    transformation: String,
    initVector: ByteArray
): ByteArray? {
    return symmetricTemplate(data, key, "AES", transformation, initVector, true)
}

/**
 * Return the bytes of symmetric encryption or decryption.
 *
 * @param data           The data.
 * @param key            The key.
 * @param algorithm      The name of algorithm.
 * @param transformation The name of the transformation, e.g., *DES/CBC/PKCS5Padding*.
 * @param isEncrypt      True to encrypt, false otherwise.
 * @return the bytes of symmetric encryption or decryption
 */
private fun symmetricTemplate(
    data: ByteArray?,
    key: ByteArray?,
    algorithm: String,
    transformation: String,
    initVector: ByteArray?,
    isEncrypt: Boolean
): ByteArray? {
    if (data == null || data.isEmpty() || key == null || key.isEmpty()) {
        return null
    }
    try {
        val secretKey: SecretKey
        if ("DES" == algorithm) {
            val desKey = DESKeySpec(key)
            val keyFactory = SecretKeyFactory.getInstance(algorithm)
            secretKey = keyFactory.generateSecret(desKey)
        } else {
            secretKey = SecretKeySpec(key, algorithm)
        }
        val cipher = Cipher.getInstance(transformation)
        if (initVector == null || initVector.isEmpty()) {
            cipher.init(if (isEncrypt) Cipher.ENCRYPT_MODE else Cipher.DECRYPT_MODE, secretKey)
        } else {
            val params = IvParameterSpec(initVector)
            cipher.init(
                if (isEncrypt) Cipher.ENCRYPT_MODE else Cipher.DECRYPT_MODE,
                secretKey,
                params
            )
        }
        return cipher.doFinal(data)
    } catch (e: Throwable) {
        e.printStackTrace()
        return null
    }
}
